torch/csrc/jit/tensorexpr/eval.cpp

```diff
@@ -672,13 +672,75 @@ class SimpleIREvaluatorImpl : public IRVisitor {
     return {};
   }
 
+  void check_bounds_throw(int64_t idx, int64_t bound, const BufPtr& buf) {
+    std::stringstream ss;
+    ss << "Index out of bounds in check_bounds. Index: " << idx
+       << "; bounds: [0, " << bound << ").";
+    throw malformed_input(ss.str(), buf);
+  }
+
+  void check_bounds(const BufPtr& buf, const std::vector<ExprPtr>& indices) {
+    const std::vector<ExprPtr>& dims = buf->dims();
+    if (dims.size() != indices.size()) {
+      // indices are flattened, but not buffer
+      if (indices.size() == 1) {
+        if (dims.size() != buf->strides().size()) {
+          throw malformed_input(
+              "Number of dimensions did not match number of strides", buf);
+        }
+        size_t buf_size = 1;
+        if (dims.size() > 0) {
+          ExprPtr buf_size_expr = immLike(dims[0], 1);
+          ExprPtr negative_one = immLike(dims[0], -1);
+          for (const auto& i : c10::irange(dims.size())) {
+            buf_size_expr = alloc<Add>(
+                buf_size_expr,
+                alloc<Mul>(
+                    alloc<Add>(dims[i], negative_one), buf->strides()[i]));
```

Comment:
Nit: not critical, but this portion might be more readable if we wrapped ExprPtrs with ExprHandles.